File: Issue-05-Scripts/scriptKmeans.py
#!/bin/env python
# -*- coding: utf-8 -*-
"""
Created on Sun Sep  6 23:14:11 2015

@author: eboza
"""
import sys, getopt, numpy
from sklearn.feature_selection import VarianceThreshold
from sklearn.cluster import KMeans
from sklearn.preprocessing import scale
from sklearn.decomposition import PCA    

def main(argv):
    ### Put the arguments in variables.
    datasetFile = ''
    clusters = 500
    try:
        opts, args = getopt.getopt(argv,"hf:k:",["datasetFile=","clusters="])
    except getopt.GetoptError:
        print 'scriptKmeans.py -f <datasetFile> -k <clusters>'
        sys.exit(2)
    for opt, arg in opts:
        if opt == '-h':
            print 'scriptKmeans.py -f <datasetFile> -k <clusters>'
            sys.exit()
        elif opt in ("-f", "--datasetFile"):
            datasetFile = arg
        elif opt in ("-k", "--clusters"):
            clusters = int(arg)
    if datasetFile == '':
        print 'ERROR: Must specify a trace filename.'
        print 'scriptKmeans.py -f <datasetFile> -k <clusters>'
        sys.exit(1)

    ### Load dataset from file passed as argument. Without the ID column.    
    dataset = numpy.loadtxt(datasetFile, delimiter=" ", skiprows=1, usecols=range(1,12))
    ### X = dataset[:,1:]
    X = dataset
    ###X_scaled = scale(X, axis=0, with_mean=False, with_std=False, copy=True)
    X_scaled = scale(X, axis=0, with_mean=True, with_std=True, copy=True)
    print "========================================\n"
    print 'Scaled Dataset Shape:'
    print(X_scaled.shape)

    ### Set similar variance for all dimensionality reduction methods:
    ### This could be read from arguments
    variance = 0.9
    
    ### Apply Variance Threshold
    selector = VarianceThreshold(threshold=variance)
    X_varianceThreshold = selector.fit_transform(X)
    print "========================================\n"
    print "Variance Threshold:\n"
    print "Shape after VarianceThreshold:" + str(X_varianceThreshold.shape) + "\n"
    print "Variance of each feature:" + str(selector.variances_) + "\n\n"

    ### Apply PCA
    pca = PCA(n_components=variance, copy=True, whiten=False)
    X_scaled_PCA = pca.fit_transform(X_scaled)
    print "========================================\n"
    print "PCA:\n"
    print "Shape after PCA:" + str(X_scaled_PCA.shape) + "\n"
    print "Variance Ratio:" + str(pca.explained_variance_ratio_) + "\n\n"

    ### K-Means with the Scaled Dataset - k=<clusters> 500 by default
    print "========================================\n"
    print "K-means only with Scaled Dataset:\n"
    ###kmeans_allFeatures = KMeans(init='k-means++', n_clusters=500, n_init=10)
    kmeans_allFeatures = KMeans(init='k-means++', n_clusters=clusters, n_init=10)
    Z_allFeatures = kmeans_allFeatures.fit_predict(X_scaled)
    numpy.savetxt(datasetFile + "_clustered-Z_allFeatures.csv", Z_allFeatures, delimiter=",", fmt="%d", header="Cluster_id", comments="")
    numpy.savetxt(datasetFile + "_clusters_centers-Z_allFeatures.csv", kmeans_allFeatures.cluster_centers_, delimiter=",")

    ### K-Means with VarianceThreshold - k=500 per elbow test
    print "========================================\n"
    print "K-means with VarianceThreshold:\n"
    if X_scaled.shape == X_varianceThreshold.shape:
        print "VarianceThreshold didn't remove any feature."
    else:
        ###kmeans_varianceThreshold = KMeans(init='k-means++', n_clusters=500, n_init=10)
        kmeans_varianceThreshold = KMeans(init='k-means++', n_clusters=clusters, n_init=10)
        Z_varianceThreshold = kmeans_varianceThreshold.fit_predict(X_varianceThreshold)
        numpy.savetxt(datasetFile + "_clustered-Z_varianceThreshold.csv", Z_varianceThreshold, delimiter=",", fmt="%d", header="Cluster_id", comments="")
        numpy.savetxt(datasetFile + "_clusters_centers-Z_varianceThreshold.csv", kmeans_varianceThreshold.cluster_centers_, delimiter=",")

    ### K-Means with PCA - k=<clusters> 500 by default
    print "========================================\n"
    print "K-means with PCA:\n"
    ###kmeans_withPCA = KMeans(init='k-means++', n_clusters=500, n_init=10)
    kmeans_withPCA = KMeans(init='k-means++', n_clusters=clusters, n_init=10)
    Z_withPCA = kmeans_withPCA.fit_predict(X_scaled_PCA)
    numpy.savetxt(datasetFile + "_clustered-Z_withPCA.csv", Z_withPCA, delimiter=",", fmt="%d", header="Cluster_id", comments="")
    numpy.savetxt(datasetFile + "_clusters_centers-Z_withPCA.csv", kmeans_withPCA.cluster_centers_, delimiter=",")

    ### K-Means 2 Features Skewness and AverageInterarrival (mean) - k=<clusters> 500 by default
    ### id mean median mid-range gmean std iqr range mad coeficiente_variacion skewness kurtosis
    print "========================================\n"
    print "K-means with 2 Features: Skewness and AverageInterarrival:\n"
    ###kmeans_twoFeatures = KMeans(init='k-means++', n_clusters=500, n_init=10)
    kmeans_twoFeatures = KMeans(init='k-means++', n_clusters=clusters, n_init=10)
    Z_twoFeatures = kmeans_twoFeatures.fit_predict(X_scaled[:,[0,9]])
    numpy.savetxt(datasetFile + "_clustered-Z_twoFeatures.csv", Z_twoFeatures, delimiter=",", fmt="%d", header="Cluster_id", comments="")
    numpy.savetxt(datasetFile + "_clusters_centers-Z_twoFeatures.csv", kmeans_twoFeatures.cluster_centers_, delimiter=",")
 
if __name__ == "__main__":
   main(sys.argv[1:])

File: Issue-05-Scripts/R/scriptKmeans.R
#### Add libraries neccesaries
for (package in c('data.table', 'getopt')) {
  if (!require(package, character.only=T, quietly=T, warn.conflicts = F)) {
    install.packages(package)
    library(package, character.only=T, quietly = T, warn.conflicts = F)
  }
}

### Functions

skscaleR <- function(dd){
  r <- scale(x = x.ds, center = TRUE, scale = apply(X = x.ds, 2, sd, na.rm = TRUE))
  return(r)
}

colVar <- function(dd){
  return (apply(dd, 2, function(x){ var(x)}))
}

varianceThreshold <- function(dd, variance){
  dd <- as.data.frame(dd)
  r <- as.matrix(dd[, colVar(dd)>= variance])
  return (r)
}

getNComponents <- function(dd, variance){
  r <-0
  cnt <- 0
  for (i in dd) {
    if(r >= variance)
      break
    r <- r + i
    cnt <- cnt + 1
  }
  return(cnt)
}

pca <- function(data, n_components = NULL){
  i <- 0
  tmp <- 0
  t.pca <- princomp(x = data)
  t.summary <- summary(t.pca)
  t.totalVariance <-sum(t.summary$sdev^2)
  t.variances <- t.summary$sdev^2 / t.totalVariance
  for(i in 1:length(t.variances)){
    tmp <- tmp + t.variances[i]
    if(tmp >= n_components)
      break;
  }
  t.transformed <-predict(t.pca, newdata = data)
  t.components <-as.data.frame(t.transformed[,1:i])
  list("pca" = t.pca, "variances" = t.variances)
}

writeResult <- function(ds, clusters, centers, filename1, filename2){
  write.table('Cluster-id', 
              file = paste(ds, filename1, sep = ''),
              row.names = F, col.names = F)
  
  write.table(clusters,
              file = paste(ds, filename1, sep = ''),
              append = T, sep = '\n', row.names = F, col.names = F)
  
  write.csv(centers,
            file = paste(ds, filename2, sep = ''),
            row.names = F)
}

### Main program

datasetFile <- ''
clusters <- 500
spec <- t(matrix(c('help', 'h', 0, "logical",
                   'file=', 'f', 2, "character",
                   'clusters=', 'k', 1, "character"), nrow = 4))

#argv <- commandArgs(trailingOnly = TRUE)
args <- c("-f","./test/Animation-lookup-100k.txt-dataset", '-k', 10)
opts = getopt(spec, args)
if(!is.null(opts$help)){
  print('scriptKmeans.R -f <datasetFile> -k <clusters>')
  quit(save = "no", status = 2)
}else if(!is.null(opts$file)){
  datasetFile <- opts$file
}
if(!is.null(opts$clusters)){
  clusters <- opts$clusters
}else{
  print('scriptKmeans.R -f <datasetFile> -k <clusters>')
  quit(save = "no", status = 2)
}
if(length(datasetFile) == 0){
  print ('ERROR: Must specify a trace filename.')
  print ('scriptKmeans.py -f <datasetFile>')
  quit(save = "no", status = 1)
}

### Load dataset from file passed as argument. Without the ID column. 
x.ds <- fread(input = datasetFile,  sep = " ", header = T, dec = ".")
x.ds$id <- NULL
x.scaled <- skscaleR(x.ds)

cat("========================================\n\n")
cat('Scaled Dataset Shape: \n')
cat(dim(x.scaled))
cat("\n")

### Set similar variance for all dimensionality reduction methods:
### This could be read from arguments
variance <- 0.9

### Apply Variance Threshold
x.varianceThreshold <- varianceThreshold(x.ds, variance)
x.varianceThreshold_var <- colVar(x.ds)
cat('========================================\n')
cat("\nVariance Threshold: \n")
cat('\nShape after Variance Threshold:\n')
cat(dim(x.varianceThreshold))
cat("\n\nVariance of each feature: \n\n")
cat(x.varianceThreshold_var)
cat("\n\n")


### Apply PCA
x_scaled_pca <- pca(data = x.scaled, n_components = variance)
cat('========================================\n\nPCA\n\n')
cat("Shape after PCA:")
d<- getNComponents(x_scaled_pca$variances, variance)
cmps <- x_scaled_pca$pca$scores[, c(1:d)]
cat(dim(cmps))
cat('\n\nVarinate Ratio: \n\n')
cat(x_scaled_pca$variances[c(1:d)])
cat("\n\n")

### K-Means with the Scaled Dataset - k=<clusters> 500 by default
cat('========================================\n\n')
cat('K-means only with Scaled Dataset:\n\n')
# In R, doesn't exist 'kmeans++'
kmeans.allFeatures = kmeans(x = x.scaled, centers = clusters, nstart = 10)

writeResult(ds = datasetFile, clusters = kmeans.allFeatures$cluster-1,
            centers = kmeans.allFeatures$centers,
            filename1 = "_clustered-Z_allFeatures.csv",
            filename2 = "_clusters_centers-Z_allFeatures.csv")

### K-Means with VarianceThreshold - k=500 per elbow test
cat('========================================\n\n')
cat('K-means with VarianceThreshold:\n\n')
if (all(dim(x.scaled) == dim(x.varianceThreshold))){
  cat('VarianceThreshold didn\'t remove any feature.')
}else{
  kmeans.varianceThreshold <- kmeans(x = x.varianceThreshold, centers = clusters, nstart = 10)

  writeResult(ds = datasetFile, clusters = kmeans.allFeatures$cluster-1,
              centers = kmeans.allFeatures$centers,
              filename1 = "_clustered-Z_varianceThreshold.csv",
              filename2 = "_clustered-centers-Z_varianceThreshold.csv")
}

### K-Means with PCA - k=<clusters> 500 by default
cat('========================================\n\n')
cat('K-means with PCA:\n\n')
###kmeans_withPCA = KMeans(init='k-means++', n_clusters=500, n_init=10)
kmeans.withPCA = kmeans(x = x_scaled_pca$pca$scores, centers = clusters, nstart = 10)

writeResult(ds = datasetFile, clusters = kmeans.withPCA$cluster-1,
            centers = kmeans.withPCA$centers,
            filename1 = "_clustered-Z_withPCA.csv",
            filename2 = "_clusters_centers-Z_withPCA.csv")

### K-Means 2 Features Skewness and AverageInterarrival (mean) - k=<clusters> 500 by default
### id mean median mid-range gmean std iqr range mad coeficiente_variacion skewness kurtosis

cat('========================================\n\n')
cat('K-means with 2 Features: Skewness and AverageInterarrival:\n\n')
kmeans.twoFeatures <- kmeans(x = x.scaled[, c(1,10)], centers = clusters, nstart = 10)

writeResult(ds = datasetFile, clusters = kmeans.twoFeatures$cluster-1,
            centers = kmeans.twoFeatures$centers,
            filename1 = "_clustered-Z_twoFeatures.csv",
            filename2 = "_clusters_centers-Z_twoFeatures.csv")
